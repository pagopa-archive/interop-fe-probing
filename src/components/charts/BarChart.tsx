import { Grid } from "@mui/material";
import { VegaLite, VisualizationSpec } from "react-vega";
import ChartsLegend from "./ChartsLegend";

const mainSpec: VisualizationSpec = {
  "title": {
    "text": "Performance dellâ€™e-service",
    "anchor": "start",
  },
  "encoding": {
    "x": {
      // connected field from the values properties
      "field": "value",
      // "quantitative" if the field is a number
      // "nominal" if the field is a string
      // "temporal" if the field is a date time object
      "type": "quantitative",
      // if the title is null it is removed
      "title": null,
      // remove the axis
      "axis": null,
    },
    "y": {
      "field": "status",
      "title": null,
      "type": "ordinal",
      "axis": null,
      // without sorting
      // the bars can be sorted by some value; by default ascending order
      "sort": null,
    },
    "color": {
      // to hide the legend
      "legend": null,
      "field": "status",
      "type": "nominal",
      "title": null,
      // color the bar parts
      "scale": {
        "domain": [
          "E-service online",
          "Monitoraggio sospeso",
          "E-service offline",
        ],
        "range": ["#17324D", "#A2ADB8", "#FE6666"],
        "type": "ordinal",
      },
    },
  },
  "config": {
    "view": {
      // hide the borders
      "stroke": "transparent",
    },
  },
  // put the values in the right of the bar rectangles as a different layer
  "layer": [
    {
      "mark": "bar",
    },
    {
      "mark": {
        "type": "text",
        "align": "left",
        "baseline": "middle",
        "dx": 3,
        "fill": "#17324D",
        "fontWeight": "bold",
        "fontSize": 15,
        "text": { "signal": "datum.value + '%'" },
      },
    },
  ],
  // height and width of the plot
  "height": 120,
  "width": 300,
};

// elements for the legend component
const legendElements = [
  { label: "E-service online", color: "#17324D" },
  { label: "Monitoraggio sospeso", color: "#A2ADB8" },
  { label: "E-service offline", color: "#FE6666" },
];

interface IService {
  status: string;
  value: number;
}

interface IProps {
  data: Array<IService>;
}

const BarChart = ({ data }: IProps) => {
  // create specification with the main spec and the value
  // and return the ready chart
  const createChart = () => {
    let spec: VisualizationSpec = {
      ...mainSpec,
      "data": {
        // type of the data, it can be "json", "csv", "tsv", "dsv"
        "values": [...data],
      },
    };
    return <VegaLite spec={spec} actions={false} />;
  };

  return (
    <Grid container direction="column" rowSpacing={2}>
      <Grid item>{createChart()}</Grid>
      <Grid item>
        <ChartsLegend legendElements={legendElements} />
      </Grid>
    </Grid>
  );
};
export default BarChart;
