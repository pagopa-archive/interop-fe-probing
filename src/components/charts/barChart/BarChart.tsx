import { max, scaleLinear, scaleBand, ScaleLinear, ScaleBand } from 'd3'
import { useTranslation } from 'react-i18next'
import { ServicePercentagesType } from '../../../types'

interface IProps {
  data: Array<ServicePercentagesType>
}
// margin convention often used with D3
const margin = { top: 20, right: 30, bottom: 20, left: 30 }
const width = 300
const height = 200

const color = ['#17324D', '#A2ADB8', '#FE6666']

/**
 * Bar chart component
 * @component
 */
export const BarChart: React.FC<IProps> = ({ data }) => {
  const { t } = useTranslation(['detailsPage'])

  const xMax: number = max(data, (d: ServicePercentagesType) => d.value) as number

  // x scale
  const x: ScaleLinear<number, number, never> = scaleLinear().domain([0, xMax]).range([0, width])

  // x scale
  const y: ScaleBand<string> = scaleBand()
    .domain(data.map((d) => d.status))
    .rangeRound([0, height - margin.left])
    .paddingInner(0.25)

  // header of the chart
  const header: JSX.Element = (
    <g className="bar-header" transform={`translate(0, ${margin.top})`}>
      <text>
        <tspan fontFamily="Titillium Web" fontSize="18px" color="#17324D" fontWeight="700">
          {t('barChartTitle')}
        </tspan>
      </text>
    </g>
  )

  // bars and bars' titles
  const bars: JSX.Element = (
    <g transform={`translate(0, ${margin.top * 2})`}>
      {data.map((d: ServicePercentagesType, i: number) => (
        <g key={d.status}>
          <rect
            className="bar"
            y={y(d.status) as number}
            width={x(d.value) as number}
            height={y.bandwidth()}
            fill={color[i % 4]}
          ></rect>
          <text
            x={x(d.value) as number}
            y={y(d.status) as number}
            fontFamily="Titillium Web"
            fontSize="24px"
            color="#17324D"
            fontWeight="600"
            transform={`translate(10, ${y.bandwidth() * 0.7})`}
          >
            {d.value + '%'}
          </text>
        </g>
      ))}
    </g>
  )

  return (
    <svg
      className="bar-chart-container"
      width={width + margin.left + margin.right}
      height={height + margin.top + margin.bottom}
      role="img"
    >
      {header}
      {bars}
    </svg>
  )
}
