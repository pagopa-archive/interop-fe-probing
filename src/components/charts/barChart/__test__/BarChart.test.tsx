import { describe, expect, test } from 'vitest'
import { render } from '@testing-library/react'
import { BarChart } from '../BarChart'

const data = [
  { status: 'E-service online', value: 90 },
  { status: 'Monitoraggio sospeso', value: 7 },
  { status: 'E-service offline', value: 15 },
]

describe('BarChart', () => {
  test('render component', () => {
    const { container } = render(<BarChart data={data} />)
    expect(container).toMatchSnapshot()
  })
})
