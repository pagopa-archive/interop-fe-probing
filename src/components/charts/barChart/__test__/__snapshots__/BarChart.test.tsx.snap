// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BarChart > render component 1`] = `
<div>
  <svg
    class="bar-chart-container"
    height="240"
    role="img"
    width="430"
  >
    <g
      class="bar-header"
      transform="translate(0, 20)"
    >
      <text>
        <tspan
          color="#17324D"
          font-family="Titillium Web"
          font-size="18px"
          font-weight="700"
        >
          barChartTitle
        </tspan>
      </text>
    </g>
    <g
      transform="translate(0, 50)"
    >
      <g>
        <rect
          class="bar"
          fill="#17324D"
          height="46"
          width="300"
          y="1"
        />
        <text
          color="#17324D"
          font-family="Titillium Web"
          font-size="24px"
          font-weight="600"
          transform="translate(10, 32.199999999999996)"
          x="300"
          y="1"
        >
          90%
        </text>
      </g>
      <g>
        <rect
          class="bar"
          fill="#A2ADB8"
          height="46"
          width="23.333333333333332"
          y="62"
        />
        <text
          color="#17324D"
          font-family="Titillium Web"
          font-size="24px"
          font-weight="600"
          transform="translate(10, 32.199999999999996)"
          x="23.333333333333332"
          y="62"
        >
          7%
        </text>
      </g>
      <g>
        <rect
          class="bar"
          fill="#FE6666"
          height="46"
          width="50"
          y="123"
        />
        <text
          color="#17324D"
          font-family="Titillium Web"
          font-size="24px"
          font-weight="600"
          transform="translate(10, 32.199999999999996)"
          x="50"
          y="123"
        >
          15%
        </text>
      </g>
    </g>
  </svg>
</div>
`;
